Added support for python3 and test script

mattpaletta opened the pull request:
I couldn't get the 'examples/run_examples.sh' script running properly, maybe something messed up with my system.  I added it in case you wanted to add automated testing for python 2 and python 3.  Great project.

mattpaletta committed:
Added support for python3 and test script

justinfx commented:
Thanks for this contribution. I haven't looked at this project in 6 years :-) Cool to see that it is still helping people. I'm surprised it still works with the ZeroMQ without having a requirements.txt file specifying the particular versions from 6 years back.

I will take a look through this merge soon!

mattpaletta commented:
Awesome, I think something is off with the package name.  Python3 doesn't pick up the imported classes properly... If you figure it out, lemme know :)

```
File "utilities/RPCService.py", line 17, in __call__
    from pyrpc import PyRpc
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pyrpc/__init__.py", line 45, in <module>
    from server import PyRpc, RpcResponse
ModuleNotFoundError: No module named 'server'
```

justinfx commented on examples/run_example.sh:
```diff
@@ -0,0 +1,14 @@
+  printf "${GREEN}Tests Passed${NC}\n"
+fi
+
+exit $TEST_EXIT_CODE
```
How does this file work correctly if the server process blocks? When would the client actually run?

mattpaletta replied:
Good catch, I posted a revision for this.  Should work correctly now.  not sure why it was working on my machine before....

justinfx replied:
This change works, but it leaves the server process running in the background. Maybe consider something like this, to clean up the server:

```
python server.py &
SERVER_PID=$!

python client.py &
CLIENT_PID=$!

wait $CLIENT_PID
TEST_EXIT_CODE=$?

kill $SERVER_PID
```

justinfx replied:
The server process is still left running after the test ends.

justinfx commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
+  </component>
+</project>
```
Are the JetBrains IDE files important to commit to the project? Shouldn't the entire .idea directory be excluded in the gitignore?

mattpaletta replied:
Sometimes part of the .idea folder is committed so developers can share workspace settings, I can ignore and remove these files if you prefer.

justinfx replied:
If there really is something of value in these project files, then maybe. But it feels weird committing files that make assumptions about IDEs. If we did that, then we might see projects with eclipse, sublime, vim, jetbrains, etc project files, accumulated over time from merge requests from different developers with different preferences.

justinfx commented:
I'm not sure how you ended up with a 'pyrpc' package name when you installed. When I install it, it uses the name 'pyRpc'. Is that still an issue for you?

mattpaletta committed:
fixed run_example.sh and added pip files to gitignore

mattpaletta commented:
Tried it on a different machine, no longer an issue.

mattpaletta commented:
Also, I need to move the __version__ out of the pyRpc package.  You can't install it without pyzmq, which setup.py tries to install. Because setup.py imports pyRpc to get the version, and it hasn't gotten dependencies yet.  I'll add this in another commit.

mattpaletta committed:
Moved version out so we can install it

mattpaletta commented:
Do you get this error in python 2.7? when running run_examples.sh


```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "build/bdist.macosx-10.6-intel/egg/pyRpc/client.py", line 199, in _worker_routine
    resp = remote.recv_pyobj()
  File "/private/tmp/pyRpc/env/lib/python2.7/site-packages/pyzmq-17.0.0-py2.7-macosx-10.6-intel.egg/zmq/sugar/socket.py", line 623, in recv_pyobj
    return self._deserialize(msg, pickle.loads)
  File "/private/tmp/pyRpc/env/lib/python2.7/site-packages/pyzmq-17.0.0-py2.7-macosx-10.6-intel.egg/zmq/sugar/socket.py", line 490, in _deserialize
    return load(recvd)
ValueError: unsupported pickle protocol: 3
```

justinfx commented:
That error is being cause by this:
http://pyzmq.readthedocs.io/en/latest/api/zmq.html#zmq.Socket.send_pyobj

Mixing python 2 and 3 won't currently work because py3 is going to default to encoding messages with protocol version 3. This doesn't exist in python 2.

In order to be cross compatible, the send_pyobj calls need to explicitly set the protocol to 2. This will ensure python 2 and 3 can communicate.

Commit:
Fixed python2/3 compatability in test and in client, and updated git ignore to remove intelliJ folders

mattpaletta commented:
Hey sorry it took so long, I added specifying the protocol in the client code.  I also added a condition to the test to make it run in python3 if available.  Since that breaks the system.  If python3 is installed, sys.version_info returns 3, but if I call `python client.py`, it still starts the python 2 interpreter (on my Mac).  Also added the git ignore for JetBrains folders.

justinfx commented on pyRpc/client.py:
```diff
@@ -163,7 +163,7 @@ def call(self, method, callback=None, async=False, args=[], kwargs={}):
             self._main_sender.connect(self._address)  
                       
-        self._main_sender.send_pyobj(req)
+        self._main_sender.send_pyobj(req, protocol=PyVersion)
```
I don't think there is any connection between the major version of Python, and the the Pickle protocol version. It just so happens that there is a 2 and a 3 right now. There is also 0 which is the ascii format, and 4 which was introduced in python 3.4. But beyond that... this would still be broken if you had a server running python3 and a client running python2. They would speak different protocols because of the python version. It is safer to just use protocol 2 across the board (I think).

mattpaletta committed:
Updated to always use version 2

justinfx commented on examples/run_example.sh:
```diff
@@ -0,0 +1,24 @@
+
+    python3 client.py &
+    wait $!
+    TEST_EXIT_CODE=$?
```
I actually had to disable this test, because my environment had both `python` and `python3` in the path, which is perfectly valid. It picked up `python3` but then tried to use zmq which was only built for python2. I guess it is just on the user to be running in a virtual env that doesn't have both python versions.

mattpaletta replied:
Good catch, Ok I'll revert to just always use `python...`

mattpaletta committed:
Kill server process and always call

justinfx commented:
Thanks for the updates. I will merge this as is, and I can fix the version in setup.py by just reading the version module in directly.

justinfx commented:
I've updated the version import in setup.py to use `execfile` so as not to trigger an import. And I have now tagged and pushed an update of 0.3.0 to pypi
